[SPARK-24225][CORE] Support closing AutoClosable objects in MemoryStore

JeetKunDoug opened the pull request:
This allows Broadcast Variables can be released properly

## What changes were proposed in this pull request?

Broadcast variables, while usually used to broadcast data to executors, can also be used to control the scope and lifecycle of shared resources (e.g. connection pools). When creating and destroying those resources within a task is expensive, using a broadcast variable to keep them deserialized in memory for multiple tasks to share can make a huge difference in the efficiency of a Spark job.

In `MemoryStore`, check if any entries in a `DeserializedMemoryEntry` implement `AutoClosable` and, if so, call `close` on those resources. This occurs in two places:
- `remove` of an individual item
- `clear` of the MemoryStore

## How was this patch tested?

Added additional tests to `MemoryStoreSuite` in order to check that we properly close resources, and handle exceptions properly.

JeetKunDoug committed:
[SPARK-24225] Support closing AutoClosable objects in MemoryStore so Broadcast Variables can be released properly

JeetKunDoug commented:
@dbtsai Here's the PR we discussed earlier.

AmplabJenkins commented:
Can one of the admins verify this patch?

dbtsai commented:
Could you change the title to `[SPARK-24225][CORE] Support closing AutoClosable objects...`? Thanks.

dbtsai commented:
Jenkins, add to whitelist.

dbtsai commented:
test this please

gatorsmile commented:
cc @jiangxb1987 @cloud-fan

SparkQA commented:
**[Test build #90599 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90599/testReport)** for PR 21322 at commit [`f254f94`](https://github.com/apache/spark/commit/f254f94fdc5e2648d7c1104bf5ec2355de7c6055).

dbtsai commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,15 +385,36 @@ private[spark] class MemoryStore(
+  private def maybeReleaseResources(entry: MemoryEntry[_]): Unit = {
+    entry match {
+      case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
+      case DeserializedMemoryEntry(objs: Array[Any], _, _) => maybeCloseValues(objs)
```
`DeserializedMemoryEntry(values, _, _)` to match the rest of code style.

dbtsai commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,15 +385,36 @@ private[spark] class MemoryStore(
+    }
+  }
+
+  private def maybeCloseValues(objs: Array[Any]): Unit = {
```
`values: Array[Any]`

dbtsai commented on core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:
```diff
@@ -526,4 +526,84 @@ class MemoryStoreSuite
+  override def close(): Unit = {
+    closed = true
+    if (throwsOnClosed) {
+      throw new RuntimeException("Throwing")
```
Could you add `var isExcpetionThrown = false`, and check it in the test whether the exception is thrown?

dbtsai commented on core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:
```diff
@@ -526,4 +526,84 @@ class MemoryStoreSuite
+    }
+  }
+  def getClosed(): Boolean = {
+    closed
```
since `closed` is public, you might use it directly. Or you can make `closed` private.

dbtsai commented on core/src/test/scala/org/apache/spark/storage/MemoryStoreSuite.scala:
```diff
@@ -526,4 +526,84 @@ class MemoryStoreSuite
+    assert(tracker.getClosed())
+  }
+
+  test("[SPARK-24225]: clear should close all AutoCloseable objects put together in an iterator") {
```
Can you check if you have multiple autocloseable objects in an iterator, and only one of them is removed, the rests of the objects will not be closed?

JeetKunDoug replied:
So if I understand the API correctly, there is no way to remove a single item that was put as part of a call to `putIterator` - because operations are conducted by `blockId` you would only be able to remove the whole group of entries, not a single part of an iterator.

dbtsai reviewed (commented):
LGTM. Only some minor comments. Let's wait more feedback from committers. Thanks.

SparkQA commented:
**[Test build #90599 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90599/testReport)** for PR 21322 at commit [`f254f94`](https://github.com/apache/spark/commit/f254f94fdc5e2648d7c1104bf5ec2355de7c6055).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90599/
Test PASSed.

viirya commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,15 +385,36 @@ private[spark] class MemoryStore(
+  private def maybeReleaseResources(entry: MemoryEntry[_]): Unit = {
+    entry match {
+      case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
+      case DeserializedMemoryEntry(objs: Array[Any], _, _) => maybeCloseValues(objs)
```
As I know, broadcasted variables can be serialized on disk too (`BlockManager.doPutIterator`). In the case, seems `AutoCloseable` broadcasted variables won't hit this release logic.

JeetKunDoug replied:
I wouldn't expect a never-deserialized Memory Entry to be closed, as it was never really instantiated to begin with - so if it _only_ lands on disk, I think that's reasonable (as the variable in question would never have had a chance to allocate anything either).

JeetKunDoug replied:
Ah- ok, I see where the issue is. So in this case you may have a deserialized instance but the memory store is full, so it fails to be put. Now we've got a live, deserialized object not in MemoryStore. Thanks for catching this. It looks like this case could be handled in `MemoryStore.putIteratorAsValues` when the `putIterator` call fails, which would handle several cases in `BlockManager` where we try (and fail) to put deserialized values, but it means a check for potential `AutoClosable` values any time we fail to put into `MemoryStore`, and I'm not sure of the performance impact of this.

JeetKunDoug replied:
Actually, digging further, there's other places where we may deserialize an object from the disk store and never put it into the memory store - it seems like punting on a guarantee that your AutoClosable object is closed and making this a best-effort thing when calling `BlockManager.removeBroadcast` (which is how we used it in the case that caused us to put together this PR) may make the most sense. It'll still be better than depending on GC and a finalizer to get the resources cleaned up when the driver can call `Broadcast#destroy` but we can document it as a best practice to have one just in case the close() call doesn't happen due to edge cases.

viirya commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,15 +385,36 @@ private[spark] class MemoryStore(
+
+  private def maybeCloseValues(objs: Array[Any]): Unit = {
+    objs.foreach {
+        case closable: AutoCloseable =>
```
indent style: two spaces.

cloud-fan commented:
I think users are responsible to call `Broadcast#destroy`, which unpersist broadcast blocks from block manager and run user-defined driver side cleanup.

It is a valid use case to allow users to define some executor side cleanup via `AutoCloseable`. However, I don't think we should always detect `AutoCloseable` when removing a block, as it may break existing program and cause perf regression. We should only do it for broadcast blocks.

A good place to do it seems to be `BlockManager.removeBroadcast`

viirya commented:
One concern I have is, now we keep broadcasted variables in `BroadcastManager.cachedValues` by using weak reference. So if a broadcasted variable with `AutoCloseable` is released before we call `Broadcast#destroy`, we still can't properly release the resources.

viirya commented:
@JeetKunDoug The same issue we discuss above. I think if there's a deserialized version of the variable, it can be not in `MemoryStore` but only serialized bytes in disk store.

The reason is, we use `getLocalValues` to retrieve the deserialized object. If it is stored in disk store by `putSingle`, we read it back and call `maybeCacheDiskValuesInMemory` trying to cache it in `MemoryStore`. But it is not guaranteed to be succeed. When it fails, you still only have serialized value in disk store.

JeetKunDoug commented:
@cloud-fan So your suggestion makes sense - it seems like the best path forward is to check the `isBroadcast` flag on the BlockId passed in to `MemoryStore.remove` and release resources only if it's a broadcast ID - does that make sense to you?

JeetKunDoug committed:
Only close `Broadcast` blocks
- Also fixed indentation issue.

SparkQA commented:
**[Test build #90651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90651/testReport)** for PR 21322 at commit [`62d46d3`](https://github.com/apache/spark/commit/62d46d3bf49ef0393a916d3cafaae4947f374f36).

JeetKunDoug committed:
Address style/variable naming issues

SparkQA commented:
**[Test build #90652 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90652/testReport)** for PR 21322 at commit [`6a08c43`](https://github.com/apache/spark/commit/6a08c434cf967b939b8065bb23d64d0715e38a2c).

JeetKunDoug commented:
@cloud-fan I added the above-mentioned check for `isBroadcast` and only release resources if it's a broadcast ID. This will affect the MapOutputTracker as well, but I think it's a pretty limited change and should work well. If you're not crazy about that, I can start at `Broadcast.destroy` and pass down a flag to control the behavior, but it's a much more invasive (but doable) change.

SparkQA commented:
**[Test build #90651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90651/testReport)** for PR 21322 at commit [`62d46d3`](https://github.com/apache/spark/commit/62d46d3bf49ef0393a916d3cafaae4947f374f36).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90651/
Test PASSed.

SparkQA commented:
**[Test build #90652 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90652/testReport)** for PR 21322 at commit [`6a08c43`](https://github.com/apache/spark/commit/6a08c434cf967b939b8065bb23d64d0715e38a2c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90652/
Test PASSed.

dbtsai commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,14 +385,37 @@ private[spark] class MemoryStore(
-        case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
-        case _ =>
+      if (blockId.isBroadcast) {
+        maybeReleaseResources(entry)
```
In this case, what happen when the blockId is not broadcast? The existing cleaning-up will not be called.

JeetKunDoug replied:
@dbtsai thanks - That whole "my day job vs. OSS" rush to fix. Will take care of it correctly and push an update.

dbtsai commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -404,6 +428,7 @@ private[spark] class MemoryStore(
 
   def clear(): Unit = memoryManager.synchronized {
     entries.synchronized {
+      entries.values().asScala.foreach(maybeReleaseResources)
```
Should we check if the keys are `blockId.isBroadcast == true`?

viirya commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,14 +385,37 @@ private[spark] class MemoryStore(
+    }
+  }
+
   def remove(blockId: BlockId): Boolean = memoryManager.synchronized {
```
To do it in `remove`, I don't think we can avoid the issue I mentioned before. If you have a deserilized value in broadcast cache, it's possible to be cleaned by GC if this broadcasted value isn't stored as deserialized entry in `MemoryStore`.

If the object already claims some resources we want to release by using `AutoCloseable` interface, we don't properly release it when it's cleaned by GC. That is happened before `remove` is called.

cloud-fan commented:
according to `TorrentBroadcast.writeBlocks` and `TorrentBroadcast.readBroadcastBlock`, broadcasted data are always written into the block manager with `MEMORY_AND_DISK_SER` mode. When a task wants to get the broadcasted value, it needs to get the serialized bytes from block manager(maybe memory, maybe disk) and deserialize it into an object, put the object into a cache, and return the object to the user function.

That said, at the executor side we may create the broadcasted object multiple times: e.g. if it's evicted from cache and GCed, we read and deserialize it from block manager again. A simple way to do executor side clean up is implementing `finalize` method in the class you want to broadcast.

Alternatively, we can improve the cache and watch the eviction event. If eviction happens, do the cleanup.

JeetKunDoug committed:
Fix handling of broadcast blocks
Needd to do the check further down the call stack so the original release still works.

JeetKunDoug commented:
@cloud-fan as to eviction, it ends up calling `BlockManager.dropFromMemory`, which will in fact call `MemoryStore.remove` and do the cleanup appropriately, so that should already be handled appropriately.

I think you're right in that the `TorrentBroadcast.readBroadcastBlock` will only succeed if it can, in fact, push the unblockified object to the MemoryStore, but I think @viirya is right that there's still a possibility that it's been spilled to disk (and appropriately closed), but then we deserialize it again in `BlockManager.getLocalValues` but fail to write it back to the MemoryStore _there_, and then it's possible we can return the value to the client but not add it to the MemoryStore. In that case, a finalize method can handle the cleanup. However, I wonder if we should always store it back to the `blockManager` in the Some() case when we read from `getLocalValues` just to ensure it's there? Since we refuse to allow it to be used if we can't save it in the case we read it from another node, why would we allow it here?

Regardless, I'm pushing some fix-ups to the other issues @dbtsai raised, which I think will handle all of the cases we'd need to consider this a "best-effort" cleanup while still requiring a finalizer. Please let me know what your thoughts on forcing the save to the memory store in `readBroadcastBlock`'s Some block where it manages to read from local values ( before TorrentBroadcast.scala:218 in my branch).

SparkQA commented:
**[Test build #90796 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90796/testReport)** for PR 21322 at commit [`790c906`](https://github.com/apache/spark/commit/790c906062fc55f553a5cecb61147f801875d4b2).

SparkQA commented:
**[Test build #90796 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90796/testReport)** for PR 21322 at commit [`790c906`](https://github.com/apache/spark/commit/790c906062fc55f553a5cecb61147f801875d4b2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90796/
Test FAILed.

kiszk commented:
retest this please

SparkQA commented:
**[Test build #90824 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90824/testReport)** for PR 21322 at commit [`790c906`](https://github.com/apache/spark/commit/790c906062fc55f553a5cecb61147f801875d4b2).

SparkQA commented:
**[Test build #90824 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/90824/testReport)** for PR 21322 at commit [`790c906`](https://github.com/apache/spark/commit/790c906062fc55f553a5cecb61147f801875d4b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/90824/
Test PASSed.

cloud-fan commented:
IIUC it's impossible to clean up broadcasted object in `MemoryStore`. The life cycle of the broadcasted object is:
1. The first task that tries to get the broadcasted object will read bytes from block manager and deserialize it to object. The object will be put into the executor-wise cache.
2. Other tasks in this executor try to get the broadcast object, read from cache if it's there.
3. Other tasks in this executor try to get the broadcast object, redo step 1 if it has been evicted from cache.
4. Job finishes, remove the value in block manager and cache.

If we do cleanup in `MemoryStore`, we just rebuild the object from bytes and call its `close`. This doesn't help as we need to do cleanup for all the objects that have been created during the job.

JeetKunDoug commented:
@cloud-fan So I see what's been messing with me - it's the whole Broadcast variable cache/week reference thing. I originally wrote this internally for Spark 1.6.3 (still supporting some older Spark installations), which didn't have this concept. I'll take another look at how we may be able to better handle the executor-wide cache in a more coherent way (pun intended) and get back to you. You may be right that we will have to depend on a combination of a finalizer + autocloseable object handled more directly in `Broadcast#destroy` for the happy-path case where we're not running into memory pressure on the MemoryStore that is causing the variable to churn.... I just hate having to depend on a finalizer to do this if there's a way around it.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

JeetKunDoug commented:
@cloud-fan (Also left on the JIRA ticket) - Sorry this has dropped off my radar for so long - work + life took me away from it for a while. So looking at the PR review comments and better understanding Broadcast Variable behavior (and some of the changes that took place in the 2.X series), it seems like simply trying to close Broadcast variables won't work as intended. However, I believe the underlying concept (driver-scoped shared variables, where the variable lives until the job is done or the driver removes it) is still worth pursuing. Being able to scope shared resources (like DB connection pools, which may need to change per phase of a job, or be able to be disposed of early in a process, which makes static variables not useful). Given that, I'd like to propose we add a new concept, similar to Broadcast Variables, called, perhaps, Scoped Variables. The intent would be for these to be scoped by the driver, be relatively small from a memory-consumption perspective (unlike broadcast variables, which can be much larger), and to be held in memory until explicitly removed by the driver. Most of the infrastructure work for broadcast variables supports this use-case, but we'd need to have either a "non-purgable" type in the MemoryStore, or some other store specific to these new scoped variables, in order to prevent them from being evicted like cached items are.
 
Thoughts on this? I'll start working on updating the PR to support something like this sometime today, but it might still take a while to get something workable put together, so I'd appreciate any feedback when someone has the time.

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,15 +385,30 @@ private[spark] class MemoryStore(
+
+  private def maybeCloseValues(values: Array[Any], blockId: BlockId): Unit = {
+    if (blockId.isBroadcast) {
+      values.foreach(value => Utils.tryClose(value))
```
Just a style thing, but could be `values.foreach(Utils.tryClose)`

srowen commented on core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala:
```diff
@@ -384,15 +385,30 @@ private[spark] class MemoryStore(
+
+  private def maybeReleaseResources(blockId: BlockId, entry: MemoryEntry[_]): Unit = {
+    entry match {
+      case SerializedMemoryEntry(buffer, _, _) => buffer.dispose()
```
Why not just make these case classes `Closeable` and then you can close them consistently

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1930,6 +1930,18 @@ private[spark] object Utils extends Logging {
     }
   }
 
+  def tryClose(value: Any): Unit = {
```
This should accept at best `AnyRef`. It doesn't really seem like we need a new global utility method for this. It's a little unusual to try closing things that aren't `Closeable` and we can try to rationalize that in the callers above if possible.

cloud-fan commented:
Introducing a new concept which is similar to broadcast seems like an overkill. We can just update broadcast, to allow it to be memory-only.

However, there might be simpler solutions to fit your case, without broadcast. e.g.
```
val myObj = ...
rdd.mapPartitions { it =>
  try {
    // process data
  } finally {
    myObj.close()
  }
}
```

I think we need to clear define the use case and think about whether we need a new API or not.

JeetKunDoug commented:
@cloud-fan The use-case we have in mind (and are currently using Broadcast + finalizers for) is the case where you have, for example, a connection pool that is, for one reason or another, scoped to a particular stage in the job. In this case, the pool itself is expensive to create, and can be shared across tasks, which makes closing the object in a try/finally for a single task, or even a single partition, as you'd end up potentially closing the resource early, and having to rebuild it several times. The fundamental trick is to figure out a way to allow the driver to define the scope of the shared resource (like a broadcast variable) and ensure it's really memory-only, so if there's a better way to use the existing broadcast variable infrastructure to do this, and prevent this kind of broadcast variable from being purged from the MemoryStore, then I'm all for it.

cloud-fan commented:
Ah i see. It's kind of a broadcast but has a much small scope and its lifecycle is bound to a stage. I'm looking forward to a full design of it, thanks!

cloud-fan commented:
I think a design doc is better, to make sure we are on the same page before the actual coding, which saves time.

JeetKunDoug commented:
Sounds good - it may take me a while to get the time to work this up, but I'll get something put together and attached to the underlying issue as soon as I can. If possible, can you point me at what you would consider a "good" design document, so I have a starting point to work with?

cloud-fan commented:
I'd like to recommend http://apache-spark-developers-list.1001551.n3.nabble.com/SPIP-SPARK-25728-Structured-Intermediate-Representation-Tungsten-IR-for-generating-Java-code-td25370.html

HyukjinKwon commented:
Shall we leave this PR closed and start it from a design doc? Let me suggest to close this for now while I am looking through old PRs.

@JeetKunDoug, please feel free to create a clone of this PR if there's any reason to keep this open that I missed. No objection.

Thanks.

JeetKunDoug commented:
@HyukjinKwon makes sense - I'll try to get a design doc put together as soon as I can, but my "day job" is preventing me from working on it right now.
